RFCs/FS-1128-better-interpolated-triple-quoted-strings.md

```diff
@@ -0,0 +1,119 @@
+- [x] Approved in principle
+- [ ] [Implementation]() (no implementation yet, but a proof of concept [here](https://github.com/dotnet/fsharp/compare/main...abonie:fsharp:poc_improved_interpolation))
+- [ ] Design Review Meeting(s) with @dsyme and others invitees
+- [Discussion]()
+
+# Summary
+
+Extend current form of interpolated triple quoted string to allow specifying how many braces should start an interpolation (equal to the number of dollar signs at the start of the literal). For example:
+```
+let json = $$"""
+{
+  "id": "Some guid",
+  "name": "{{user.Name}}",
+};
+"""
+```
+In the above example, `$$` starting the string literal indicates that interpolated expression will be enclosed in `{{...}}` and single `{` and `}` will be treated as content.
+
+# Motivation
+
+Triple quoted strings are already quite handy for embedding e.g. XML in string literals, but not as convenient for embedding some other languages that use `{` and `}` in their syntax, e.g. JSON or CSS. Currently every occurrence of a curly brace in an interpolated string needs to be escaped, which can quickly become tedious. With this extended syntax, only the braces around interpolated expression need special attention, and that is where programmer's focus already is.
+
+The motivation is not to enable any novel use cases, but to make working with interpolated string literals a much more seamless experience in the use cases mentioned above.
+
+This change should:
+1. be backwards compatible with existing support for triple quoted strings
+2. align with C# raw string literals reasonably well
+
+# Detailed design
+
+The existing form of interpolated triple quoted strings: `$"""...{}..."""` is extended to allow multiple dollar signs at the begining of the string literal. The count of these initiatory `$` characters indicates how many `{` and `}` characters are used to delimit interpolation expression within the content of the literal.
+
+Important note: To maintain backward compatibility, there are no changes in semantics for a single-`$` case. This means that it is an edge case in terms of how this feature behaves. It would be nice to have a clean design that covers all cases, but it is not worth breaking backward compatibility for. Therefore the below description applies to the literals starting with 2 or more `$`.
+
+Triple quoted string literals starting with 2 or more `$` characters have no escaping mechanism for `{` or `}`. Instead, the literal can always be constructed in such a way as to ensure that interpolation delimiters will not collide with other curly braces in the content. In a literal that starts with `N` `$` characters, a sequence of consecutive `{` can be at most `2*N-1` characters long and for any sequence of more than `N`, the innermost `N` `{` characters delimit the interpolation, while the outter up to `N-1` are treated as content (analogously for `}`). A sequence of `2*N` or more `{` (or `}`) will result in a compilation error.
```

Comment:
Can you break these long single line paragraph sentences up by adding line breaks? It’ll remain rendered as a single oars, but will help with adding targeted comments, and is easier on the diffs and people editing or viewing the source.

I needed to read this sentence three times. Though I don’t immediately have a clear simpler version in mind. Maybe this came from the C# spec?

Perhaps we could add a bunch of smaller examples with more of the variants shown and how they relate to 2N-1.

Also, is there a limit? I.e., what does the compiler do when encountering, say, 100 leading $ signs?

Reply:
In C#, even 111 of dollar signs still works:

https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEUCuA7AHwAEAmARgFgAoANwEMoACAWwE8A1BgSzuABsYAOTrMYjALyMArAG5G1ekzYBlDFC54A5hMYASfQcNHjJ02fMXLV68YBE9gCoALLgGdGblq0YYYCDADeQcEhoWHhEZFR0TGxcWFsnOq8AsKiAL6ZWdk5uXn5BYVFxSWlufa2MtTURGQAnAAUKmoamgCUMkA===

Reply:
There's no hardcoded limit, and I don't think there's a limit mentioned in C# documentation. Maxint amount of dollar signs would definitely be problematic. I think Scaling section is meant for such details, but I haven't filled it out yet.